2단계 미션 제출합니다

iimpala opened the pull request:
지난 PR 코멘트 반영해서 작업 완료했습니다 :)

findstar commented on src/main/java/model/Request.java:
```diff
@@ -0,0 +1,92 @@
+        return parameter.get(key);
+    }
+
+    public Map<String, String> getAllParameter(){
```
굳이 getAllParameter 로 Map<String, String> 을 반환한다면
일급 컬렉션 객체로써의 Parameter 객체의 의미가 퇴색됩니다.
User 객체를 생성하기 위해서라면
User 객체에 Parameter 를 인자로 하는 static 메서드를 추가하는 방법도 있습니다.

findstar commented on src/main/java/model/Request.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    public boolean isLogin(){
+        String cookie= Optional.ofNullable(requestHeader.get("Cookie")).orElse("logined=false");
+        return cookie.equals("logined=true");
```
```java
cookie.equals("logined=true");
``` 
위와 같이 로그인 여부를 판단하면 나중에 
```java 
loginged=true&param1=value1
``` 
와 같이 여러개의 값을 판단할 수 없습니다.

findstar commented on src/main/java/controller/CreateUserController.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    void doPost(Request request, Response response) {
+        ObjectMapper mapper = new ObjectMapper();
+        User user = mapper.convertValue(request.getAllParameter(), User.class);
+        DataBase.addUser(user);
+        logger.info("UserInfo: {}, {}, {}, {}", user.getUserId(), user.getEmail(), user.getName(), user.getPassword());
+        response.sendRedirect("/index.html");
```
1. 회원가입시 아무런 입력값 없이 `회원가입` 버튼을 누르면 에러가 발생합니다.
 예외사항에 대한 처리로직을 추가해주세요.
2. DataBase.addUser 는 직접 DAO 에 접근하는것과 같습니다. 별도로 분리해보면 어떨까요? 
3. ObjectMapper 는 편의를 위해서 사용하신건가요? 좀 더 명시적으로 작성하는 것이 나아보입니다.

findstar commented on src/main/java/controller/LoginController.java:
```diff
@@ -0,0 +1,37 @@
+        ObjectMapper mapper = new ObjectMapper();
+        User loginUser = mapper.convertValue(request.getAllParameter(), User.class);
+        User user= Optional.ofNullable(DataBase.findUserById(loginUser.getUserId()))
+                .orElseThrow(NullPointerException::new);
```
만약 데이터베이스에서 해당 User 를 찾지 못한 경우에 반환하는 exception 이 NullPointException 일까요? 
NotFoundUserException 정도가 더 적당할것 같네요.

findstar reviewed (changes requested):
안녕하세요. mark 웹서버 리뷰를 맡은 제임스입니다. 😁
전체적으로 요구사항을 충족하는 기능 구현을 잘 해주셨네요!
몇가지 개선되면 좋을 부분에 코멘트 달아두었습니다. 
확인하시고 궁금한 점은 언제든지 DM 주세요 😄 
그럼 화이팅입니다!

limdef commented:
안녕하세요 제임스!

먼저 리뷰 감사드립니다.
피드백 주신 부분 최대한 반영해보았습니다..!
혹시 놓친 부분이 있거나, 더 개선하면 좋겠다 싶은 부분 피드백 주시면 감사하겠습니다.

findstar commented on src/main/java/controller/CreateUserController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Override
+    void doPost(Request request, Response response) {
```
Input 값에 대한 validation 처리가 추가되면 좋겠습니다 😄

findstar commented on src/test/java/webserver/HttpResponseTest.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private OutputStream createOutputStream(String filename) throws FileNotFoundException {
+        return new FileOutputStream(new File(testDirectory + filename));
```
ByteArrayStream 을 사용하시면 굳이 fileOutputStream 을 사용하지 않아도 테스트 할 수 있습니다.

findstar reviewed (approved):
안녕하세요. 리뷰어 제임스입니다.
피드백 잘 반영해주셨네요. 👏 
PR 머지하겠습니다. 
몇가지 추가적으로 남김 코멘트는 참고만 해주세요 😄 
미션 수고하셨습니다.

iimpala committed:
refactor(HandlerMapping): Controller 관련 구조 리팩토링

- Controller 인터페이스, AbstractController 추상 클래스를 작성하여 uri별로 컨트롤러 분리
- HandlerMapping클래스를 작성하여 uri와 controller를 매핑

iimpala committed:
refactor(ContentType): ContentType enum 타입 구현

ContentType을 enum객체로 구조화

iimpala committed:
feat(UserLoginService): 사용자 로그인 기능 구현 및 테스트 코드 작성

사용자 로그인 기능 구현

iimpala committed:
refactor(http): 패키지 구조 변경

request, response 패키지를 http 패키지 하위로 이동

iimpala committed:
style(RequestHandler): 불필요한 로그 제거\

iimpala committed:
feat(HttpCookie): HttpCookie 클래스 작성 및 로그인 프로세스 개선

- HttpCookie 객체를 만들어 Request/Response의 쿠키값을 관리
- 로그인 이후 JSESSIONID를 만들어 Set-Cookie헤더에 전달

iimpala committed:
feat(HttpCookie): 쿠키 구조 변경

쿠키에 name-value 이외의 속성을 추가하기 위해 쿠키 구조 변경
- HttpCookie는 기존에 모든 쿠키를 담고있는 형태에서 쿠키 1개에 대한 객체로 용도 변경
- 모든 쿠키에 대한 정보는 request, response 객체에 List로 보관

iimpala committed:
feat(list.html): list.html 수정

템플릿 엔진을 사용하기 위해 html 파일 수정

iimpala committed:
feat(UserListController): UserListController 작성

템플릿 엔진을 사용하여 사용자 목록을 동적으로 렌더링

iimpala committed:
feat(HttpSession): 세션 기반 로그인 구현

- 로그인시 세션에 사용자 정보를 저장
- 로그인한 사용자가 /user/login에 접근시 index.html로 리다이렉트
- 로그인하지 않은 사용자가 /user/login에 접근시 /user/login.html로 리다이렉트

sah3122 commented on src/main/java/webserver/HandlerMapping.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class HandlerMapping {
```
handlerMapping 클래스 정의 잘해주셨습니다 👍

sah3122 commented on src/main/java/webserver/HandlerMapping.java:
```diff
@@ -0,0 +1,43 @@
+        Optional<Controller> optionalController = handlerMapping.keySet().stream()
+                .filter(path::equals)
+                .map(handlerMapping::get)
+                .findFirst();
```
`handlerMapping.get()`으로 구현하지 않은 이유가 있을까요 ?

iimpala replied:
처음에는 `key ->  path.startsWith(key)`로 필터링을 하다가 `equals`로 바꾸면서 이렇게 남았는데, 그러면 `get()`이나 `getOrDefault()`를 써도 동일한 결과를 얻을 수 있겠네요!
리뷰 감사합니다 :)

sah3122 commented on src/main/java/webserver/controller/AbstractController.java:
```diff
@@ -0,0 +1,27 @@
+import webserver.http.response.HttpResponse;
+import webserver.http.response.HttpResponseFactory;
+
+public class AbstractController implements Controller{
```
AbstractController 정의 👍

sah3122 commented on src/main/java/webserver/controller/AbstractController.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    protected HttpResponse doGet(HttpRequest request) {
+        return HttpResponseFactory.badRequest();
```
잘못 정의된 케이스에 대해 badRequest 응답을 주도록 구현해주셨네요 👍 
다른 방법으론 Exception을 던지는것도 선호합니다 😄

iimpala replied:
컨트롤러에 대한 예외처리 부분이 부족한 것 같아 AbstractController에서 예외를 잡아 BadRequest 응답으로 변환하도록 반영하였습니다.
리뷰 감사합니다 :)

sah3122 commented on src/main/java/webserver/controller/ResourceController.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.util.Optional;
+
+public class ResourceController extends AbstractController {
```
정적 리소스를 담당하는 클래스 정의 👍

sah3122 commented on src/main/java/webserver/controller/UserCreateController.java:
```diff
@@ -0,0 +1,45 @@
+        return saveUser(request, parser.parse(body));
+    }
+
+    private HttpResponse saveUser(HttpRequest request, Map<String, String> param) {
```
save 와 같은 비즈니스 로직을 메소드로 분리해주셨는데, 개인적으론 비즈니스를 담당하는 코드만 존재하고 HttpResponse에 관한 정보는 분리하는것을 선호합니다 👍

sah3122 commented on src/main/java/webserver/controller/UserListController.java:
```diff
@@ -0,0 +1,49 @@
+        Optional<String> logined = request.getCookie().stream()
+                .filter(cookie -> "logined".equals(cookie.getName()))
+                .map(HttpCookie::getValue)
+                .findFirst();
```
코드가 복잡한것 같아요. 
`request.getCookie().get("logined")`와 같이 간결하게 변경하는건 어떨까요 ?

추가로, 로그인 여부는 Cookie를 통해 검증하기 보단 Session을 통해 검증하는것이 어떨까요 ? 
두 객체의 차이점을 고민해보세요 😄

iimpala replied:
요구사항에 적혀있어서 일단 그대로 했는데, 저도 세션을 통해 로그인 여부를 검증하는게 더 좋다고 생각합니다:)
리뷰 감사합니다 😊

sah3122 commented on src/main/java/webserver/controller/UserLoginController.java:
```diff
@@ -0,0 +1,85 @@
+
+        if (jSessionId.isEmpty()) {
+            return HttpResponseFactory.redirect(request, "/user/login.html");
+        }
```
Cookie의 `jSessionId`로 유효성 검사를 하면 어떤 문제점이 있을까요 ?

iimpala replied:
이 코드의 의도 자체는 

1. request의 "JSESSIONID"라는 이름의 쿠키가 없을 경우 리다이렉트
2. 이후 "JSESSIONID" 쿠키의 값을 키로 가지는 세션이 없을 경우 리다이렉트

이렇게 이중으로 리다이렉션을 처리해주기 위함이었습니다.

쿠키의 값 자체를 사용하여 검증을 한다면 위변조 가능성이 있어 문제가 생길 수 있지만,
현재 방식은 세션을 조회하기 전에 쿠키의 유무를 검증하는 것이라고 생각해서 문제는 없다고 생각합니다.

현재는 request의 쿠키를 Map의 형태로 관리하도록 변경하고, 키가 없으면 value="" 인 쿠키를 만들어 반환하도록 변경하여 세션 존재여부를 검증할 때 리다이렉션되도록 변경했습니다 :)

sah3122 replied:
상세한 설명 감사합니다 😄

sah3122 commented on src/main/java/webserver/http/HttpCookie.java:
```diff
@@ -0,0 +1,49 @@
+package webserver.http;
+
+public class HttpCookie {
```
쿠키 클래스 정의 잘해주셨습니다 👍

sah3122 commented on src/main/java/webserver/http/response/HttpResponse.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class HttpResponse {
```
HttpResponse 정의 잘해주셨네요 👍

sah3122 reviewed (changes requested):
2단계 미션 구현 잘해주셨습니다 👍 
몇가지 고민거리를 남겨두었는데 확인 부탁드려요 🙏

iimpala committed:
refactor(HandlerMApping): getController 로직 변경

불필요한 Optional, stream 로직 제거

iimpala committed:
refactor(UserCreateController): saveUser 로직 수정

saveUser메소드에서 Http 의존성 제거

iimpala committed:
refactor(HttpRequest): request cookie 자료구조 변경 및 로그인 검증 방식 변경

- 기존 List를 통해 쿠키를 보관하던 방식에서 Map으로 변경하여 사용성을 높임
- 로그인을 기존 "logined"쿠키의 값을 통해 확인하던 방식에서 세션의 유무를 통해 확인하는 방식으로 변경

iimpala committed:
refactor(AbstractController): controller에서 예외 발생시 기본적으로 400 Bad Request를 응답하도록 변경

Controller의 예외처리를 위해 AbstractController에 예외 응답 변환 로직을 추가

iimpala commented:
피드백 주신 내용 보완했습니다!

sah3122 commented on src/main/java/webserver/controller/UserCreateController.java:
```diff
@@ -18,7 +18,8 @@ public UserCreateController(UserService userService) {
     @Override
     protected HttpResponse doGet(HttpRequest request) {
-        return saveUser(request, request.getQueryParams());
+        saveUser(request.getQueryParams());
```
피드백 반영 잘해주셨네요 👍

sah3122 commented on src/main/java/webserver/controller/UserListController.java:
```diff
@@ -27,12 +30,11 @@ public UserListController(Handlebars handlebars, UserService userService) {
-                .map(HttpCookie::getValue)
-                .findFirst();
+        HttpCookie cookie = request.getCookie().get("JSESSIONID");
 
-        if (logined.isEmpty() || "false".equals(logined.get())) {
+        String jSessionId = cookie.getName();
+        Optional<HttpSession> session = sessionManager.getSession(jSessionId);
+        if (session.isEmpty()) {
```
session 값으로 로그인 유무를 판단할수 있게 변경 잘해주셨습니다 😄

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
2단계 미션까지 잘 구현해주셔서 여기서 마무리 짓도록할게요 🙏 
남은 미션들은 필수가 아닌 선택이므로 미션을 진행하신다면 리뷰요청 부탁드립니다 !
